odoo/addons/base/models/ir_ui_view.py

```diff
@@ -580,62 +577,84 @@ def _get_filter_xmlid_query(self):
-        assert from_clause == '"ir_ui_view"'
-        self.flush(['active'])
-        query = """
+        sub_where_clause = where_clause.replace('ir_ui_view', 'view')
+        assert from_clause == '"ir_ui_view"', f"Unexpected from clause: {from_clause}"
+
+        self._flush_search(domain, fields=['inherit_id', 'priority', 'model', 'mode'], order='id')
+        query = f"""
             WITH RECURSIVE ir_ui_view_inherits AS (
-                SELECT id, inherit_id, priority
+                SELECT id, inherit_id, priority, mode, model
                 FROM ir_ui_view
-                WHERE inherit_id = %s AND {where_clause}
+                WHERE id IN %s AND {where_clause}
             UNION
-                SELECT iuv.id, iuv.inherit_id, iuv.priority
-                FROM ir_ui_view iuv
-                INNER JOIN ir_ui_view_inherits iuvi ON iuvi.id = iuv.inherit_id
-                WHERE {sub_where_clause}
+                SELECT view.id, view.inherit_id, view.priority, view.mode, view.model
+                FROM ir_ui_view view
+                INNER JOIN ir_ui_view_inherits parent ON parent.id = view.inherit_id
+                WHERE coalesce(view.model, '') = coalesce(parent.model, '')
+                      AND view.mode = 'extension'
+                      AND {sub_where_clause}
             )
-            SELECT id
-            FROM ir_ui_view_inherits
-            ORDER BY priority, id;
-        """.format(
-            where_clause=where_clause,
-            sub_where_clause=where_clause.replace('ir_ui_view', 'iuv'),
-        )
-        self.env.cr.execute(query, [self.id] + where_params + where_params)
-        view_ids = [r[0] for r in self.env.cr.fetchall()]
+            SELECT
+                v.id, v.inherit_id,
+                ARRAY(SELECT r.group_id FROM ir_ui_view_group_rel r WHERE r.view_id=v.id)
+            FROM ir_ui_view_inherits v
+            ORDER BY v.priority, v.id
+        """
+        # ORDER BY v.priority, v.id:
+        # 1/ sort by priority: abritrary value set by developers on some
+        #    views to solve "dependency hell" problems and force a view
+        #    to be combined earlier or later. e.g. all views created via
+        #    studio have a priority=99 to be loaded last.
+        # 2/ sort by view id: the order the views were inserted in the
+        #    database. e.g. base views are placed before stock ones.
+
+        self.env.cr.execute(query, [tuple(self.ids)] + where_params + where_params)
+        rows = self.env.cr.fetchall()
+
+        # filter out forbidden views
+        if any(row[2] for row in rows):
+            user_groups = set(self.env.user.groups_id.ids)
+            rows = [row for row in rows if not (row[2] and user_groups.isdisjoint(row[2]))]
 
+        views = self.browse(row[0] for row in rows)
+
+        # optimization: fill in cache of inherit_id
+        self.env.cache.update(views, type(self).inherit_id, [row[1] for row in rows])
```

Comment:
Maybe we should also fill in the cache of `mode`, since that field will be used in `_combine()`.